BUG: use more generic type inference for fast plotting

jreback opened the pull request:
xref #15073

https://travis-ci.org/pandas-dev/pandas/jobs/190356982

was failing on tz-aware dtypes. these routines are the same in that they require datetime64ns (but any *also* accepts tz-aware).

jreback committed:
BUG: use more generic type inference for fast plotting

xref #15073

jreback commented:
@jorisvandenbossche

jorisvandenbossche commented on pandas/tseries/converter.py:
```diff
@@ -172,7 +172,7 @@ def _dt_to_float_ordinal(dt):
     is a :func:`float`.
     """
     if (isinstance(dt, (np.ndarray, Index, Series)
-                   ) and is_datetime64_ns_dtype(dt)):
```
Theoretically, this is not fully correct, as `is_datetime64_any_dtype` can also accept datetime64 values with other than ns resolution? 
Although I don't think you can end up here with such values when going through the DatetimeConverter, as datetime64 arrays first get passed through `to_datetime`

jorisvandenbossche replied:
We could maybe add a test for this (and for tz aware Timestamp): https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/tests/test_converter.py#L58

jorisvandenbossche replied:
Actually, we already have this:

```
        rs = self.dtc.convert(np.array([
            np_datetime64_compat('2012-01-01 00:00:00+0000'),
            np_datetime64_compat('2012-01-02 00:00:00+0000')]), None, None)
        self.assertEqual(rs[0], xp)
```

does already test this as the above has [s] resolution, not [ns] .

Would just add a test case for the tz case.

jorisvandenbossche reviewed (commented):
Good catch! 
Forgot to look at the 'allowed failures' ..

jreback commented:
ok, I fixed ``is_datetime64_ns_dtype`` to work with tz-aware (as _any_ does as well, this makes sense). let's see if anything else breaks. added a test as well.

jreback commented:
ok, this is actually a tiny public API change, but it was inconsistent / wrong before.

jreback commented:
@jorisvandenbossche

jorisvandenbossche commented on pandas/tseries/tests/test_converter.py:
```diff
@@ -62,6 +62,21 @@ def test_conversion(self):
+        self.assertEqual(rs[1], xp)
+
+        rs = self.dtc.convert(ts, None, None)
+        self.assertEqual(rs, xp)
```
This seems duplicate of line 71-72 ?

jorisvandenbossche replied:
Can you also add  similar test for `Index([ts - Day(1), ts]).to_pydatetime` to test tz aware datetime.datetime objects (as this is actually what was raising the tests I think, as `_mpl_repr()` does return that).

jorisvandenbossche commented on pandas/tests/types/test_inference.py:
```diff
@@ -805,6 +809,33 @@ def test_is_float(self):
+        self.assertTrue(is_datetime64_any_dtype('datetime64[ns]'))
+        self.assertTrue(is_datetime64_any_dtype(ts))
+        self.assertTrue(is_datetime64_any_dtype(tsa))
+
```
Can you do the exact same 4 tests for `is_datetime64tz_dtype`?

jorisvandenbossche commented:
How safe is it to change this? Couldn't there be places in our code where `is_datetime64_ns_dtype` is that assumes it has no timezone?

Just as overview, is this correct?
- `is_datetime64_dtype` -> any resolution, no tz (pure numpy)
- `is_datetime64_any_dtype` -> any resolution, with or without tz
- `is_datetime64tz_dtype` -> ns? resolution, with tz (as other resolution with tz is not possible I think?)
- `is_datetime64_ns_dtype` -> ns resolution, with or without tz

The combination missing is "ns resolution, no tz (pure numpy)", which is actually what `is_datetime64_ns_dtype` was before. But maybe we don't need this? (as you can use the tz one to check it has *no* timezone)

jreback commented:
these functions are a bit redundant; the ``_ns_`` and ``_any_`` are barely used at all. so I don't think this is a big deal to change.

jorisvandenbossche commented:
Indeed, my search says that `_any_` is exactly used in one spot, and `_ns_` in three :-). So maybe rather something to clean-up? (but not for this PR)

jreback commented:
@jorisvandenbossche yeah I'd leave them for now, they *are* useful, even if not used :>

pushing in a sec

jreback commented:
@jorisvandenbossche

jorisvandenbossche commented:
yes!
